### PR TITLE
Implement input/ouput groups in data model

### PR DESCRIPTION
**Pull request by henhuy**
Now input and output groups can be defined as lists in output results.
See `tests/_files/industry_example.csv` for example upload file

### CODE REVIEW

File: django_comparison_dashboard/models.py [hunk 1]
```diff
@@ -30,6 +30,8 @@ class ScalarData(models.Model):
     new = models.BooleanField()
     unit = models.CharField(max_length=255)
     groups = ArrayField(models.CharField(max_length=255), null=True)
```
Comment (bereba) [inline django_comparison_dashboard/models.py:32]: Should we now name these process_groups?

File: django_comparison_dashboard/datamodel/datapackage.json [hunk 2]
```diff
@@ -41,6 +41,8 @@
         {"name": "groups", "description": "Free classification with key-value pairs", "type": "array", "unit": null},
         {"name": "input_commodity", "description": "It describes any type of energy or energy carrier (e.g. electricity, heat, solar radiation, natural gas, ...) that enters a technology.", "type": "string", "unit": null},
         {"name": "output_commodity", "description": "It describes any type of energy or energy carrier (e.g. electricity, heat, hydrogen, LNG, CO2, ...) that exits a technology.", "type": "string", "unit": null},
+        {"name": "input_groups", "description": "Free classification with key-value pairs", "type": "array", "unit": null},
```
Comment (bereba) [inline django_comparison_dashboard/datamodel/datapackage.json:44]: Why does the description say "key-value" pairs, although we give the values in a list?

**Review by bereba — COMMENTED** [review]
Thank you @henhuy. Looks good, just two minor remarks/questions.

**Comment by bereba** [comment]
Another aspect to consider: at the moment all rows have to tagged with groups if they shell be plotted in "group mode".
I think for our application, it might be better to be able to group only selected processes/commodities.
A pragmatic way to do so could be to first set e.g. input_group = input_commodity and afterward overwrite input_group with the rows have entries?

See change from upper screenshot to this one (apart from harmonizing group tags as described in previous comment):

image

**Comment by henhuy** [comment]
> I tested the new functionality with another input file. It's nice but I think writing multiple group tags into one group leads to only one group consisting of both tags. Which can be problematic if as only those are grouped which always have the same two tags. See on the uploaded screenshot for "primary/fossil" and "primary" coming from the imports. Therefore, I think for the sankey there should be only one grouping using the current functionality, or am I missing something?
> 
> image

You have to filter input/ouput groups first - so that you either have only primary tag or both...
It's a feature, not a bug ;)

File: django_comparison_dashboard/datamodel/datapackage.json (hunk 2 above)
Comment (henhuy) [reply django_comparison_dashboard/datamodel/datapackage.json:44]: Don't know where this comes from - I just copy-pasted it. will adapt

**Comment by henhuy** [comment]
> Another aspect to consider: at the moment all rows have to tagged with groups if they shell be plotted in "group mode". I think for our application, it might be better to be able to group only selected processes/commodities. A pragmatic way to do so could be to first set e.g. input_group = input_commodity and afterward overwrite input_group with the rows have entries?
> 
> See change from upper screenshot to this one (apart from harmonizing group tags as described in previous comment):
> 
> image

Maybe we should only use process,input and output groups instead of process/input_commodities/output_commodities - then we can get rid of those second columns.
But there is no possibility to group some commodities and left others untouched - but this is the case for both approaches

**Comment by bereba** [comment]
> > I tested the new functionality with another input file. It's nice but I think writing multiple group tags into one group leads to only one group consisting of both tags. Which can be problematic if as only those are grouped which always have the same two tags. See on the uploaded screenshot for "primary/fossil" and "primary" coming from the imports. Therefore, I think for the sankey there should be only one grouping using the current functionality, or am I missing something?
> > image
> 
> You have to filter input/ouput groups first - so that you either have only primary tag or both... It's a feature, not a bug ;)

Ah okay I see :) It requires no adjustments, then, but for our analysis, I'll just start with one group tag per column for the sankeys to make the usage easier. I think multiple group tags normally might be more relevant for distinct plots.

**Comment by bereba** [comment]
> > Another aspect to consider: at the moment all rows have to tagged with groups if they shell be plotted in "group mode". I think for our application, it might be better to be able to group only selected processes/commodities. A pragmatic way to do so could be to first set e.g. input_group = input_commodity and afterward overwrite input_group with the rows have entries?
> > See change from upper screenshot to this one (apart from harmonizing group tags as described in previous comment):
> > image
> 
> Maybe we should only use process,input and output groups instead of process/input_commodities/output_commodities - then we can get rid of those second columns. But there is no possibility to group some commodities and left others untouched - but this is the case for both approaches

I don't understand the first suggestion. Then we would loose some information, wouldn't we?

Couldn't the approach I suggested be implemented in the output data adapter?
Something like "if process_group.value is empty copy process.value".And then plot process_groups. 
I think then some groups would be applied and others untouched?

**Comment by henhuy** [comment]
> > > Another aspect to consider: at the moment all rows have to tagged with groups if they shell be plotted in "group mode". I think for our application, it might be better to be able to group only selected processes/commodities. A pragmatic way to do so could be to first set e.g. input_group = input_commodity and afterward overwrite input_group with the rows have entries?
> > > See change from upper screenshot to this one (apart from harmonizing group tags as described in previous comment):
> > > image
> > 
> > 
> > Maybe we should only use process,input and output groups instead of process/input_commodities/output_commodities - then we can get rid of those second columns. But there is no possibility to group some commodities and left others untouched - but this is the case for both approaches
> 
> I don't understand the first suggestion. Then we would loose some information, wouldn't we?
> 
> Couldn't the approach I suggested be implemented in the output data adapter? Something like "if process_group.value is empty copy process.value".And then plot process_groups. I think then some groups would be applied and others untouched?

I thought of input_group holding input_comodity as first tag, like:
`input_groups = ["coal", "primary", "fossile"]`
Then in sankey, you could decide which detail level you want to use.
Column `input_commodity` could be deleted then (same procedure for other columns)

**Pull request merged**